akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream.{ Attributes, FlowShape, Inlet, NeverMaterializedException, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val pr = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
+            initializing.onFuture(tr)
+          case None =>
+            val cb = getAsyncCallback(initializing.onFuture)
+            future.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!pr.isCompleted) {
+          pr.failure(new NeverMaterializedException())
+        }
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          //just leave it in the port
+          println("initialized: pushed")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onPull(): Unit = pull(in)
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            pr.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
```

Comment:
not sure I'm following here, can u please elaborate?

Reply:
I think @jrudolph may be suggesting moving the logic of `onFuture` (and calling it `connect`?) back up to enclosing `GraphStageLogic` since you're interacting with its members and not those of the `initializing` type.